File: pkg/resources/builtin_extensions/batch.py
#!/usr/bin/env python3
# ---
# name: batch
# description: Execute multiple tools and synthesise their outputs ephemerally
# builtin: true
# commands: batch: Run tools and synthesise results (e.g. /batch read 3 files and summarise)
# ---
"""batch.py — ephemeral multi-tool orchestration via the fir extension SDK.

Provides a `/batch` slash command and a `batch_run` tool that let the agent
(or user) execute a list of tool calls, collect their outputs without
polluting the main conversation, and synthesise the results via a one-shot
LLM call.

This is the sole implementation of batch/multi-tool orchestration in fir.

Architecture:
  1. Parse the tool list from the request.
  2. Execute each tool sequentially via ``ctx.call_tool()``.
     - Results are held in local Python memory — never enter history.
  3. Build a synthesis prompt from collected outputs + user instructions.
  4. Run ``ctx.side_query()`` to get an ephemeral LLM summary.
  5. Return only the summary.
"""

from __future__ import annotations

import fir_ext

# ---------------------------------------------------------------------------
# Helper: extract text from a tool result
# ---------------------------------------------------------------------------


def _result_text(result: dict) -> str:
    """Extract text content from a call_tool result dict."""
    content = result.get("content", [])
    if isinstance(content, list):
        parts = []
        for block in content:
            if isinstance(block, dict):
                text = block.get("text") or block.get("Text", "")
                if text:
                    parts.append(text)
            elif isinstance(block, str):
                parts.append(block)
        return "\n".join(parts)
    if isinstance(content, str):
        return content
    return str(result)


def _build_synthesis_prompt(
    results: list[dict],
    instructions: str,
) -> str:
    """Build the prompt sent to side_query() for synthesis."""
    parts = [
        "You are processing the outputs of multiple tool calls. "
        "Below are the results, followed by instructions on what "
        "to return.\n"
    ]
    for i, r in enumerate(results, 1):
        name = r["name"]
        error_tag = " [ERROR]" if r.get("is_error") else ""
        parts.append(f"--- Tool {i}: {name}{error_tag} ---")
        parts.append(r["output"])
        parts.append("")
    parts.append("--- Instructions ---")
    parts.append(instructions)
    return "\n".join(parts)


# ---------------------------------------------------------------------------
# Core: run a batch of tool calls and synthesise
# ---------------------------------------------------------------------------


def _run_batch(
    tools: list[dict],
    instructions: str,
    ctx: fir_ext.Context,
    description: str = "",
) -> dict:
    """Execute *tools*, collect outputs, synthesise via side_query().

    Parameters
    ----------
    tools : list of dict
        Each entry has ``"name"`` (str) and optional ``"params"`` (dict).
    instructions : str
        Synthesis instructions for the LLM.
    ctx : fir_ext.Context
        Extension context for call_tool / side_query.
    description : str, optional
        Human-readable label for progress messages.

    Returns
    -------
    dict
        Structured tool result with ``content`` and ``is_error``.
    """
    if not tools:
        return _error("tools list is empty")
    if not instructions:
        return _error("instructions are required")

    # Validate tool names and params upfront.
    available = ctx.list_tools()
    tool_index = {t["name"]: t for t in available}
    available_names = sorted(tool_index.keys())

    errors = []
    for i, spec in enumerate(tools, 1):
        name = spec.get("name", "")
        if not name:
            errors.append(f"tools[{i}]: name is required")
            continue
        if name not in tool_index:
            errors.append(
                f"tools[{i}]: tool {name!r} not found. "
                f"Available: {', '.join(available_names)}"
            )
            continue
        # Validate required params against schema.
        schema = tool_index[name].get("parameters") or {}
        required = schema.get("required") or []
        params = spec.get("params") or {}
        missing = [r for r in required if r not in params]
        if missing:
            errors.append(
                f"tools[{i}] ({name}): missing required params: "
                + ", ".join(missing)
            )

    if errors:
        return _error("Validation failed:\n" + "\n".join(errors))

    results: list[dict] = []

    for spec in tools:
        name = spec["name"]
        params = spec.get("params") or {}

        # Call the tool via the bridge.
        try:
            result = ctx.call_tool(name, params)
        except Exception as exc:
            results.append({
                "name": name,
                "output": f"error calling tool: {exc}",
                "is_error": True,
            })
            continue

        is_error = result.get("is_error", False)
        output = _result_text(result)
        results.append({
            "name": name,
            "output": output,
            "is_error": is_error,
        })

    # Synthesise collected outputs.
    prompt = _build_synthesis_prompt(results, instructions)
    try:
        synthesis = ctx.side_query(prompt)
    except Exception as exc:
        return _error(f"synthesis failed: {exc}")

    return {
        "content": [{"type": "text", "text": synthesis}],
        "is_error": False,
    }


def _error(msg: str) -> dict:
    return {
        "content": [{"type": "text", "text": msg}],
        "is_error": True,
    }


# ---------------------------------------------------------------------------
# Tool: batch_run
# ---------------------------------------------------------------------------


@fir_ext.tool(
    name="batch_run",
    description=(
        "Execute multiple tools and synthesise their outputs via an "
        "ephemeral LLM call. Raw tool outputs never enter conversation "
        "history — only the synthesis is returned.\n\n"
        "Use when you need to gather data from several tools and want "
        "a concise summary without bloating context."
    ),
    parameters={
        "type": "object",
        "properties": {
            "description": {
                "type": "string",
                "description": "Brief label for this batch (shown in UI).",
            },
            "tools": {
                "type": "array",
                "description": "Ordered list of tool calls.",
                "items": {
                    "type": "object",
                    "properties": {
                        "name": {
                            "type": "string",
                            "description": "Name of the tool to call. Use tool names from this agent's tool set, not provider built-in tool names.",
                        },
                        "params": {
                            "type": "object",
                            "description": "Tool parameters.",
                        },
                    },
                    "required": ["name"],
                },
            },
            "instructions": {
                "type": "string",
                "description": (
                    "Instructions for the LLM that synthesises "
                    "collected outputs."
                ),
            },
        },
        "required": ["tools", "instructions"],
    },
)
def batch_run(params: dict, ctx: fir_ext.Context):
    tools = params.get("tools", [])
    instructions = params.get("instructions", "")
    description = params.get("description", "")
    return _run_batch(tools, instructions, ctx, description)


# ---------------------------------------------------------------------------
# Command: /batch
# ---------------------------------------------------------------------------


@fir_ext.command(
    name="batch",
    description=(
        "Run tools and synthesise results. "
        "Usage: /batch <natural language description of what to do>"
    ),
)
def cmd_batch(args: list[str], ctx: fir_ext.Context):
    """Handle /batch — pass the request to the agent as a user message
    instructing it to use the batch_run tool."""
    text = " ".join(args).strip()
    if not text:
        return {
            "message": (
                "Usage: /batch <description>\n"
                "Example: /batch read the 5 largest .go files "
                "and summarise their purpose"
            ),
        }
    prompt = (
        f"Use the batch_run tool to accomplish the following. "
        f"Build the appropriate tool list and instructions, "
        f"then call batch_run:\n\n{text}"
    )
    ctx.send_user_message(prompt)
    return {}


fir_ext.run(name="batch")

File: pkg/agent/tools/batch.go
package tools

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/kfet/fir/pkg/agent"
	"github.com/kfet/fir/pkg/ai"
)

// BatchToolProvider is the interface the batch tool needs from the session.
// It provides access to the current tool set for dispatching sub-tool calls
// and to SimplePrompt for synthesising gathered output.
type BatchToolProvider interface {
	// GetTools returns the current tool set. The batch tool uses this to
	// look up and execute sub-tools by name.
	GetTools() *agent.ToolSet
	// SimplePrompt makes a one-shot LLM call with the given messages.
	// Used to synthesise collected tool outputs into a summary.
	SimplePrompt(ctx context.Context, messages []agent.AgentMessage) (string, error)
}

// NewBatchTool creates the batch tool. It executes a list of tool calls,
// collects their output, and pipes it through a SimplePrompt with
// user-provided instructions. Only the synthesis result is returned to the
// caller — raw tool outputs are ephemeral.
func NewBatchTool(provider BatchToolProvider) agent.AgentTool {
	return agent.AgentTool{
		Tool: ai.Tool{
			Name: "batch",
			Description: "Execute multiple tools and synthesise their outputs via a one-shot LLM call. " +
				"Raw tool outputs are ephemeral — only the LLM synthesis is returned to you. " +
				"Use this when you need to gather large amounts of data from several tools " +
				"and want a concise summary without polluting your context window.\n\n" +
				"The tool list is executed sequentially. Each tool's output is collected and " +
				"then passed along with your instructions to a single LLM call that produces " +
				"the final result.",
			Parameters: batchSchema,
		},
		Label: "batch",
		Execute: func(ctx context.Context, toolCallID string, params map[string]any, onUpdate agent.AgentToolUpdateCallback) (agent.AgentToolResult, error) {
			return executeBatch(ctx, provider, params, onUpdate)
		},
	}
}

var batchSchema = map[string]any{
	"type": "object",
	"properties": map[string]any{
		"description": map[string]any{
			"type":        "string",
			"description": "Brief description of what this batch is doing (shown in UI).",
		},
		"tools": map[string]any{
			"type":        "array",
			"description": "Ordered list of tool invocations to execute. Each entry names a tool and provides its parameters.",
			"items": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"name": map[string]any{
						"type":        "string",
						"description": "Name of the tool to call.",
					},
					"params": map[string]any{
						"type":        "object",
						"description": "Parameters to pass to the tool.",
					},
				},
				"required": []string{"name"},
			},
		},
		"instructions": map[string]any{
			"type":        "string",
			"description": "Instructions for the LLM that will process the collected tool outputs. Describe what to extract, summarise, or return.",
		},
	},
	"required": []string{"tools", "instructions"},
}

// batchToolCall is a parsed tool invocation from the batch params.
type batchToolCall struct {
	Name   string         `json:"name"`
	Params map[string]any `json:"params"`
}

// batchToolResult holds the outcome of a single sub-tool execution.
type batchToolResult struct {
	Name    string
	Index   int
	Output  string
	IsError bool
}

func executeBatch(
	ctx context.Context,
	provider BatchToolProvider,
	params map[string]any,
	onUpdate agent.AgentToolUpdateCallback,
) (agent.AgentToolResult, error) {
	// Parse tool calls.
	calls, err := parseBatchToolCalls(params)
	if err != nil {
		return errorResult(err.Error()), nil
	}
	if len(calls) == 0 {
		return errorResult("tools array is empty"), nil
	}

	instructions, _ := params["instructions"].(string)
	if instructions == "" {
		return errorResult("instructions is required"), nil
	}

	description, _ := params["description"].(string)

	tools := provider.GetTools()
	if tools == nil {
		return errorResult("no tools available"), nil
	}

	// Execute each tool sequentially, collecting results.
	results := make([]batchToolResult, 0, len(calls))
	for i, call := range calls {
		// Check context before each tool.
		if ctx.Err() != nil {
			return errorResult("batch cancelled: " + ctx.Err().Error()), nil
		}

		// Report progress.
		if onUpdate != nil {
			progress := fmt.Sprintf("[batch] executing %d/%d: %s", i+1, len(calls), call.Name)
			if description != "" {
				progress = fmt.Sprintf("[batch: %s] executing %d/%d: %s", description, i+1, len(calls), call.Name)
			}
			onUpdate(agent.AgentToolResult{
				Content: []ai.ToolResultContent{{Type: "text", Text: progress}},
			})
		}

		result := executeSubTool(ctx, tools, call, i)
		results = append(results, result)
	}

	// Build the synthesis prompt from collected outputs.
	synthesisPrompt := buildSynthesisPrompt(results, instructions)

	// Report synthesis phase.
	if onUpdate != nil {
		progress := "[batch] synthesising results..."
		if description != "" {
			progress = fmt.Sprintf("[batch: %s] synthesising results...", description)
		}
		onUpdate(agent.AgentToolResult{
			Content: []ai.ToolResultContent{{Type: "text", Text: progress}},
		})
	}

	// Run SimplePrompt to synthesise.
	msgs := []agent.AgentMessage{
		agent.NewAgentMessage(ai.NewUserMsg(synthesisPrompt, time.Now().UnixMilli())),
	}

	synthesis, err := provider.SimplePrompt(ctx, msgs)
	if err != nil {
		return errorResult("synthesis failed: " + err.Error()), nil
	}

	return agent.AgentToolResult{
		Content: []ai.ToolResultContent{{Type: "text", Text: synthesis}},
	}, nil
}

// executeSubTool looks up a tool by name and executes it.
func executeSubTool(ctx context.Context, tools *agent.ToolSet, call batchToolCall, index int) batchToolResult {
	tool, found := tools.Get(call.Name)
	if !found {
		return batchToolResult{
			Name:    call.Name,
			Index:   index,
			Output:  fmt.Sprintf("tool %q not found", call.Name),
			IsError: true,
		}
	}

	if tool.Execute == nil {
		return batchToolResult{
			Name:    call.Name,
			Index:   index,
			Output:  fmt.Sprintf("tool %q has no execute function", call.Name),
			IsError: true,
		}
	}

	// Don't allow batch to call itself (prevent infinite recursion).
	if call.Name == "batch" {
		return batchToolResult{
			Name:    call.Name,
			Index:   index,
			Output:  "batch cannot call itself",
			IsError: true,
		}
	}

	toolParams := call.Params
	if toolParams == nil {
		toolParams = make(map[string]any)
	}

	result, err := tool.Execute(ctx, fmt.Sprintf("batch-%d", index), toolParams, nil)
	if err != nil {
		return batchToolResult{
			Name:    call.Name,
			Index:   index,
			Output:  err.Error(),
			IsError: true,
		}
	}

	// Collect text content from the tool result.
	var sb strings.Builder
	for _, c := range result.Content {
		if c.Type == "text" && c.Text != "" {
			if sb.Len() > 0 {
				sb.WriteString("\n")
			}
			sb.WriteString(c.Text)
		}
	}

	return batchToolResult{
		Name:    call.Name,
		Index:   index,
		Output:  sb.String(),
		IsError: result.IsError,
	}
}

// buildSynthesisPrompt constructs the prompt for the SimplePrompt call.
func buildSynthesisPrompt(results []batchToolResult, instructions string) string {
	var sb strings.Builder
	sb.WriteString("You are processing the outputs of multiple tool calls. ")
	sb.WriteString("Below are the results, followed by instructions on what to return.\n\n")

	for _, r := range results {
		sb.WriteString(fmt.Sprintf("--- Tool %d: %s", r.Index+1, r.Name))
		if r.IsError {
			sb.WriteString(" [ERROR]")
		}
		sb.WriteString(" ---\n")
		sb.WriteString(r.Output)
		sb.WriteString("\n\n")
	}

	sb.WriteString("--- Instructions ---\n")
	sb.WriteString(instructions)
	sb.WriteString("\n")

	return sb.String()
}

// parseBatchToolCalls extracts the tool call list from batch params.
func parseBatchToolCalls(params map[string]any) ([]batchToolCall, error) {
	rawTools, ok := params["tools"]
	if !ok {
		return nil, fmt.Errorf("tools is required")
	}

	arr, ok := rawTools.([]any)
	if !ok {
		return nil, fmt.Errorf("tools must be an array")
	}

	calls := make([]batchToolCall, 0, len(arr))
	for i, item := range arr {
		obj, ok := item.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("tools[%d]: must be an object", i)
		}

		name, _ := obj["name"].(string)
		if name == "" {
			return nil, fmt.Errorf("tools[%d]: name is required", i)
		}

		var toolParams map[string]any
		if p, ok := obj["params"]; ok {
			if pm, ok := p.(map[string]any); ok {
				toolParams = pm
			}
		}

		calls = append(calls, batchToolCall{Name: name, Params: toolParams})
	}

	return calls, nil
}

func errorResult(msg string) agent.AgentToolResult {
	return agent.AgentToolResult{
		Content: []ai.ToolResultContent{{Type: "text", Text: msg}},
		IsError: true,
	}
}
